src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs

```diff
@@ -59,36 +59,57 @@ public async Task<IEnumerable<EmissionsData>> GetCarbonIntensityAsync(IEnumerabl
         }
         return result;
     }
-    
+
     /// <inheritdoc />
     public async Task<EmissionsForecast> GetCurrentCarbonIntensityForecastAsync(Location location)
     {
         this.Logger.LogInformation($"Getting carbon intensity forecast for location {location}");
+        using (var activity = Activity.StartActivity())
+        {
+            var balancingAuthority = await this.GetBalancingAuthority(location, activity);
+                var forecast = await this.WattTimeClient.GetCurrentForecastAsync(balancingAuthority); 
+                return ForecastToEmissionsForecast(forecast, location);
+        } 
+    }
 
+    /// <inheritdoc />
+    public async Task<EmissionsForecast> GetCarbonIntensityForecastAsync(Location location, DateTimeOffset generatedAt)
+    {
+        this.Logger.LogInformation($"Getting carbon intensity forecast for location {location} requested at {generatedAt}");
         using (var activity = Activity.StartActivity())
         {
-            BalancingAuthority balancingAuthority = await this.GetBalancingAuthority(location, activity);
-            var data = await this.WattTimeClient.GetCurrentForecastAsync(balancingAuthority);
-
-            var duration = GetDurationFromGridEmissionDataPoints(data.ForecastData.FirstOrDefault(), data.ForecastData.Skip(1)?.FirstOrDefault());
-            
-            // Linq statement to convert WattTime forecast data into EmissionsData for the CarbonAware SDK.
-            var forecastData = data.ForecastData.Select(e => new EmissionsData() 
-            { 
-                Location = e.BalancingAuthorityAbbreviation, 
-                Rating = ConvertMoerToGramsPerKilowattHour(e.Value), 
-                Time = e.PointTime,
-                Duration = duration
-            });
-
-            return new EmissionsForecast()
+            var balancingAuthority = await this.GetBalancingAuthority(location, activity);
+            var forecast = await this.WattTimeClient.GetForecastOnDateAsync(balancingAuthority, (DateTimeOffset)generatedAt!);
+            if (forecast == null)
             {
-                GeneratedAt = data.GeneratedAt,
-                Location = location,
-                ForecastData = forecastData,
-            };
+                Exception ex = new ArgumentException($"No forecast was generated at the requested time {generatedAt}");
+                activity?.SetStatus(ActivityStatusCode.Error, ex.Message);
+                Logger.LogError(ex, ex.Message);
```

Comment:
Yeah we went back and forth on this a bit but it was essentially how to inform the user that the forecast they requested doesn't actually exist. Seems like the user would like to know, and this was what we thought was the best interaction. Other options were
1. return a null value in the list of results. DIdn't seem ideal. 
2. return an object with null values/empty forecast data field. But then it's unclear if the user simply filtered out all the forecast data or there actually wasn't a forecast to start with

Reply:
Totally makes sense.  I just want to double check that an error for the user is also an error for the operator here.

The example from my past that makes me critical of this is `if (user.password != input.password) then CreateErrorResponse() & LogError()`  Error to the user, valid system response.  No need to wake up an on-call engineer ;)

I agree with the user flow.  The operator experience is more nuanced so I wanted to ask about it.  Since we expect WattTime to return `400` if there aren't any forecasts, if we hit this `200` path case we would want an operator to know and investigate what was going on.

Thanks for the details.  I think what's here is great as is :)